ext/nmatrix/storage/list/list.cpp

```diff
@@ -602,6 +603,45 @@ void init_default(LIST_STORAGE* s) {
   s->default_val = NM_ALLOC(D);
   *reinterpret_cast<D*>(s->default_val) = 0;
 }
+  
+template <typename DType>
+bool is_symmetric(const LIST_STORAGE* s) {
+  
+  NODE *i_curr, *j_curr;
+  size_t count = 0;
+  std::map<std::pair<int, int>, VALUE> my_dict;
+  std::pair<int, int> key;
+  VALUE value;
+ 
+  if (s->dim != 2) {
+    rb_raise(rb_eNotImpError, "is_symmetric only defined for dim = 2");
+  }
+
+  for (i_curr = s->rows->first; i_curr; i_curr = i_curr->next) {
+    int i = i_curr->key - s->offset[0];
+    if (i < 0 || i >= (int)s->shape[0]) continue;
+
+    for (j_curr = ((LIST*)(i_curr->val))->first; j_curr; j_curr = j_curr->next) {
+      int j = j_curr->key - s->offset[1];
+      if (j < 0 || j >= (int)s->shape[1]) continue;
+      if (i != j) {
+        value = *((VALUE*)j_curr->val);
+        key = std::make_pair(i, j);
+        if( my_dict.count(key) == 1 ) {
+            if (my_dict[key] != value) {
+              return false;
+            }
+        }
+        else {
+            key = std::make_pair(j, i);
+            my_dict.insert(std::make_pair(key,value));
+        }
+      }
```

Comment:
This is my first attempt to implement symmetric for list storage type. Please take a look and provide some feedback.

It iterates over all nodes (all non default elements) and uses a dictionary to see whether value of row i and col j is similar to value of row j and col i and hence find whether the matrix is symmetric or not.

The complexity is nlogn where n is number of non default elements in the matrix. The complexity can be reduced to n by using unordered_map instead of map. The reason I haven't done that is because unordered_set cannot take two integers as key. It can be done by defining separate hash function. Should I do that?

What do you think?